ably/realtime_channel.go

```diff
@@ -383,32 +421,41 @@ func (em ChannelEventEmitter) OffAll() {
 	em.emitter.OffAll()
 }
 
-// Publish publishes a message on the channel, which is send on separate
-// goroutine. Publish does not block.
+// Publish publishes a message on the channel.
 //
-// This implicitly attaches the channel if it's not already attached.
-func (c *RealtimeChannel) Publish(name string, data interface{}) (Result, error) {
-	return c.PublishAll([]*proto.Message{{Name: name, Data: data}})
+// This implicitly attaches the channel if it's not already attached. If the
+// context is canceled before the attach operation finishes, the call
+// returns with an error, but the operation carries on in the background and
+// the channel may eventually be attached and the message published anyway.
+func (c *RealtimeChannel) Publish(ctx context.Context, name string, data interface{}) error {
+	return c.PublishAll(ctx, []*Message{{Name: name, Data: data}})
 }
 
 // PublishAll publishes all given messages on the channel at once.
-// PublishAll does not block.
 //
-// This implicitly attaches the channel if it's not already attached.
-func (c *RealtimeChannel) PublishAll(messages []*proto.Message) (Result, error) {
+// This implicitly attaches the channel if it's not already attached. If the
+// context is canceled before the attach operation finishes, the call
+// returns with an error, but the operation carries on in the background and
+// the channel may eventually be attached and the message published anyway.
+func (c *RealtimeChannel) PublishAll(ctx context.Context, messages []*Message) error {
```

Comment:
I think this should be called `PublishBatch` as it's publishing a batch of messages (and we use the term "batch" in the spec when referring to the publish of multiple messages).

Reply:
Sounds better. Done at 8f632a1.

Reply:
This name is being superseded by https://github.com/ably/ably-go/pull/303. @lmars PTAL